src/Microsoft.Identity.Client/Logger.cs

```diff
@@ -0,0 +1,199 @@
+﻿//------------------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using System.Globalization;
+using Microsoft.Identity.Client.Internal;
+
+namespace Microsoft.Identity.Client
+{
+    public sealed class Logger
+    {
+        public enum LogLevel
+        {
+            Error = 0,
+            Warning = 1,
+            Info = 2,
+            Verbose = 3
+        }
+
+        internal Logger(Guid correlationId)
+        {
+            this.CorrelationId = correlationId;
+        }
+
+        internal Guid CorrelationId { get; set; }
+
+        /// <summary>
+        /// Callback instance
+        /// </summary>
+        /// 
+        private static readonly object LockObj = new object();
+        private static ILoggerCallBack _localCallback;
+
+        public static ILoggerCallBack Callback
+        {
+            set
+            {
+                lock (LockObj)
+                {
+                    if (_localCallback != null)
+                    {
+                        throw new System.Exception("MSAL logging callback can only be set once per process and" +
```

Comment:
>throw [](start = 24, length = 5)

this is not completely thread safe. Refer to [https://msdn.microsoft.com/en-us/library/ff650316.aspx](https://msdn.microsoft.com/en-us/library/ff650316.aspx).
if you make the callback volatile then code should be like...
set{
            lock (LockObj) 
            {
               if (instance != null) 
                 //fail
            }
//set value
}